std/experimental/allocator/package.d

```diff
@@ -549,16 +928,18 @@ to be shared across threads, use $(D processAllocator) (below). By default,
 $(D theAllocator) ultimately fetches memory from $(D processAllocator), which
 in turn uses the garbage collected heap.
 */
-nothrow @safe @nogc @property IAllocator theAllocator()
+nothrow @safe @nogc
+@property ref RCIAllocator theAllocator()
 {
-    auto p = _threadAllocator;
-    return p !is null ? p : setupThreadAllocator();
+    alias p = _threadAllocator;
+    return !p.isNull() ? p : setupThreadAllocator();
 }
 
 /// Ditto
-nothrow @safe @nogc @property void theAllocator(IAllocator a)
+nothrow @system @nogc
+@property void theAllocator(RCIAllocator a)
```

Comment:
Is it defined whether allocators can always free their backing memory on destruction, or whether they have to leak when there are still outstanding allocations?
I'm assuming the former as the bookkeeping overhead for the latter seems problematic.

If that's the case, making the `theAllocator` setter unsafe is quite a severe choice.
Sure you correctly force code that replaces the current allocator to be unsafe, but how could that code take any responsibility for all the allocations that were done with the previous `theAllocator` (or at least for that code to be reasonably auditable).

This would limit the safe Allocator usages to:
- setting `theAllocator` once on startup
- using local allocators for specific tasks
- others???
But excluded (b/c crazily unsafe) those usages:
- exchanging `theAllocator` after some time to optimize for the current workload
- others???

Global allocators with finite lifetime are inherently unsafe. We should have a good answer to how we intend allocators to be used, before making this step.

As a side node, the old `void theAllocator(IAllocator a)` was already unsafe, as it could eliminate the last GC reference to the previous `theAllocator`. Allocations don't GC-pin their allocator and aren't usually GC tracked anyhow.

On a lighter note, this is the second special RC type (after [RC Throwables](https://github.com/dlang/dmd/pull/6681)), that's ending up in dlang before we're set on how to implement manual memory management in the language. Another class-based RC type instead of a polymorphic RC!(IAllocator). It's good to experiment, but it remains questionable whether std.experimental is the right place ;).
Anyhow, hopefully this won't cause too much maintenance effort down the road.

Reply:
>But excluded (b/c crazily unsafe) those usages:
> - exchanging `theAllocator` after some time to optimize for the current workload

What if the user wraps his allocator in a `StatsCollector` allocator that tracks the `bytesUsed`? This way he could safely change `theAllocator`

I'm thinking of something like:

```
alias SCAlloc = StatsCollector!(Mallocator, Options.bytesUsed);
SCAlloc statsCollectorAlloc;

// setup the allocator
() @trusted { theAllocator = allocatorObject(statsCollectorAlloc); }();
// do some work

// check if we can safely change the allocator
if (theAllocator.impl.bytesUsed == 0) { // change theAllocator }
```

Reply:
Nice, good idea.
Haven't thought too well about how that would interact with possible compile-time lifetime checks (for RC et.al.), but keeping track of the outstanding bytes should allow for a cheap runtime safety check before exchanging `theAllocator`.
Would it be relevant whether the allocator set as `theAllocator` has been used beforehand? Would this be worthwhile to add as a static introspectable `empty` method to the allocator interfaces?

For `processAllocator` keeping track of bytes atomically might be somewhat expensive. Some clever per-thread accounting/caching could lower the overhead there.

Reply:
>Would it be relevant whether the allocator set as theAllocator has been used beforehand? 

Good question. The design is young, and at some point we'll need to answer this.

>Would this be worthwhile to add as a static introspectable empty method to the allocator interfaces?

We already have `empty`: https://dlang.org/phobos/std_experimental_allocator.html#.IAllocator.empty

>For processAllocator keeping track of bytes atomically might be somewhat expensive. Some clever per-thread accounting/caching could lower the overhead there.

This is tricky because the very notion of "empty" is evanescent for a shared allocator. You can't check for `empty` and then change the allocator - that's a race. We'd need some CAS-based mechanism. Let's discuss this further as needs arise.